homeassistant/components/proximity/coordinator.py

```diff
@@ -0,0 +1,246 @@
+from homeassistant.util.location import distance
+from homeassistant.util.unit_conversion import DistanceConverter
+
+from .const import (
+    ATTR_DIR_OF_TRAVEL,
+    ATTR_DIST_TO,
+    ATTR_NEAREST,
+    CONF_IGNORED_ZONES,
+    CONF_TOLERANCE,
+    DEFAULT_DIR_OF_TRAVEL,
+    DEFAULT_DIST_TO_ZONE,
+    DEFAULT_NEAREST,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ProximityDataUpdateCoordinator(
+    DataUpdateCoordinator[dict[str, str | int | float]]
+):
+    """Proximity data update coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, friendly_name: str, config: ConfigType
+    ) -> None:
+        """Initialize the Proximity coordinator."""
+        self.ignored_zones: list[str] = config[CONF_IGNORED_ZONES]
+        self.proximity_devices: list[str] = config[CONF_DEVICES]
+        self.tolerance: int = config[CONF_TOLERANCE]
+        self.proximity_zone: str = config[CONF_ZONE]
+        self.unit_of_measurement: str = config.get(
+            CONF_UNIT_OF_MEASUREMENT, hass.config.units.length_unit
+        )
+        self.friendly_name = friendly_name
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=friendly_name,
+            update_interval=None,
+        )
+
+        self.data = {
+            ATTR_DIST_TO: DEFAULT_DIST_TO_ZONE,
+            ATTR_DIR_OF_TRAVEL: DEFAULT_DIR_OF_TRAVEL,
+            ATTR_NEAREST: DEFAULT_NEAREST,
+        }
+
+        self.old_state: State | None = None
+        self.new_state: State | None = None
+        self.entity: str
+        async_track_state_change(
+            hass, self.proximity_devices, self.async_check_proximity_state_change
+        )
+
+    async def async_check_proximity_state_change(
+        self, entity: str, old_state: State | None, new_state: State | None
+    ) -> None:
+        """Fetch and process state change event."""
+        self.old_state = old_state
+        self.new_state = new_state
+        self.entity = entity
+        await self.async_refresh()
+
+    async def _async_update_data(self) -> dict[str, Any]:
+        """Calculate Proximity data."""
+        data: dict[str, str | int | float] = {
+            ATTR_DIST_TO: DEFAULT_DIST_TO_ZONE,
+            ATTR_DIR_OF_TRAVEL: DEFAULT_DIR_OF_TRAVEL,
+            ATTR_NEAREST: DEFAULT_NEAREST,
+        }
+
+        if self.new_state is None:
+            _LOGGER.debug("no new_state -> abort")
+            return self.data
+
+        entity_name = self.new_state.name
+        devices_to_calculate = False
+        devices_in_zone = ""
+
+        zone_state = self.hass.states.get(f"zone.{self.proximity_zone}")
+        proximity_latitude = (
+            zone_state.attributes.get(ATTR_LATITUDE) if zone_state else None
+        )
+        proximity_longitude = (
+            zone_state.attributes.get(ATTR_LONGITUDE) if zone_state else None
+        )
+
+        # Check for devices in the monitored zone.
+        for device in self.proximity_devices:
```

Comment:
Are these devices or entities? Given its pulling `hass.states` i'm assuming they are entities, so wondering if the variables should be renamed, or if there is something else here specific to home assistant devices.

I realize this is moving existing code so it could also be future cleanup.

Reply:
you are right, these are actually entities of type `person` or `device_tracker` ... but as you mentioned, this is moving existing code, therefore i tried to change only the needed parts and leave the rest as it is